pageserver/src/page_cache.rs

```diff
@@ -226,51 +254,290 @@ pub struct CacheEntryContent {
+
+						// reconstruct most recent page version
+						if content.wal_record.is_some() {
+							// force reconstruction of most recent page version
+							self.reconstruct_page(key, content)?;
+						}
+
+						maxbuf.clear();
+						maxkey.pack(&mut maxbuf);
+
+						if last_lsn > horizon {
+							// locate most recent record before horizon
+							let mut iter = self.db.iterator(IteratorMode::From(&maxbuf[..], Direction::Reverse));
+							if let Some((k,v)) = iter.next() {
+								minbuf.clear();
+								minbuf.extend_from_slice(&v);
+								let content = CacheEntryContent::unpack(&mut minbuf);
+								if content.wal_record.is_some() {
+									minbuf.clear();
+									minbuf.extend_from_slice(&k);
+									let key = CacheKey::unpack(&mut minbuf);
+									self.reconstruct_page(key, content)?;
+								}
+							}
+						}
+						// remove records prior to horizon
+						minbuf.clear();
+						minkey.pack(&mut minbuf);
+						self.db.delete_range_cf(cf, &minbuf[..], &maxbuf[..])?;
+
+						maxkey = minkey;
+					}
+				}
+			}
+		}
+	}
+
+	fn reconstruct_page(&self, key: CacheKey, content: CacheEntryContent) -> anyhow::Result<Bytes> {
+        let entry_rc = Arc::new(CacheEntry::new(key.clone(), content));
+
+        let mut entry_content = entry_rc.content.lock().unwrap();
+        entry_content.apply_pending = true;
+
+        let s = &self.walredo_sender;
+        s.send(entry_rc.clone())?;
+
+        while entry_content.apply_pending {
+            entry_content = entry_rc.walredo_condvar.wait(entry_content).unwrap();
+        }
+        // We should now have a page image. If we don't, it means that WAL redo
+        // failed to reconstruct it. WAL redo should've logged that error already.
+        let page_img = match &entry_content.page_image {
+            Some(p) => p.clone(),
+            None => {
+                error!(
+                    "could not apply WAL to reconstruct page image for GetPage@LSN request"
+                );
+                bail!("could not apply WAL to reconstruct page image");
+            }
+        };
+        self.put_page_image(key.tag, key.lsn, page_img.clone());
+		Ok(page_img)
+	}
+
+	fn wait_lsn(&self, lsn: u64) -> anyhow::Result<()> {
+		let mut shared = self.shared.lock().unwrap();
+		let mut waited = false;
+
+        // There is a a race at postgres instance start
+        // when we request a page before walsender established connection
+        // and was able to stream the page. Just don't wait and return what we have.
+        // TODO is there any corner case when this is incorrect?
+        if !shared.walreceiver_works {
+            trace!(
+                " walreceiver doesn't work yet last_valid_lsn {}, requested {}",
+                shared.last_valid_lsn,
+                lsn
+            );
+        }
+
+        if shared.walreceiver_works {
+
+			while lsn > shared.last_valid_lsn {
+				// TODO: Wait for the WAL receiver to catch up
+				waited = true;
```

Comment:
All the intents here are wrong. Please consider setting up your editor to automatically use the standard rustfmt style? There should be lots of guides online for how to do this.